### PR TITLE
 implemented selecting specific month ref: #46424

### PR DESCRIPTION
**Pull request by TheWalkingLeek**

### CODE REVIEW

**Comment by psunix** [comment]
latest-screenshot

validation error when date_from before date_to

**Comment by psunix** [comment]
latest-screenshot

when I enter an invalid year, ember crashes

File: frontend/app/templates/components/date-from-to-edit.hbs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  <thead>
+    <tr>
+      <td scope="col"></td>
+      <td class="text-center" scope="col">Monat</td>
+      <td class="text-center" scope="col">Jahr</td>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th class="align-middle" scope="row">
```
Comment (psunix) [inline frontend/app/templates/components/date-from-to-edit.hbs]: create own component or helper month-year-input to make this code more DRY

File: app/exporters/odt/cv.rb [hunk 2]
```diff
@@ -110,12 +110,12 @@ def project_description(project)
     end
 
     def formatted_year(obj)
-      if obj.year_to.nil?
-        "#{obj.year_from} - heute"
-      elsif obj.year_from == obj.year_to
-        obj.year_to
+      if obj.date_to.nil?
```
Comment (psunix) [inline app/exporters/odt/cv.rb]: is it possible to call these attrs: start_at, finish_at ?
would be more rails like

File: frontend/app/components/date-from-to-edit.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  },
+
+  setYear(year, option) {
+    let date = this.getDateObj(option)
+    date.setFullYear(year)
+    this.set('entity.date_' + option.toLowerCase(), date);
+    this.set('year' + option, year);
+  },
+
+  getDateObj(option) {
```
Comment (psunix) [inline frontend/app/components/date-from-to-edit.js]: other name for method and param ?
maybe you could move this to the model class ? computed property

File: .gitignore [hunk 4]
```diff
@@ -86,3 +86,6 @@ fabric.properties
 .project
 
 package-lock.json
+
+# Coverage
+/spec/coverage
```
[the remark this answers is not in the document]
Comment (psunix) [reply .gitignore]: these changes are still here ...

File: app/models/project.rb [hunk 5]
```diff
@@ -25,13 +25,12 @@ class Project < ApplicationRecord
 
   has_many :project_technologies, dependent: :destroy
 
-  validates :year_from, :person_id, :role, :title, presence: true
-  validates :year_from, :year_to, length: { is: 4 }, allow_blank: true
+  validates :finish_at, :person_id, :role, :title, presence: true
   validates :description, :technology, :role, length: { maximum: 5000 }
   validates :title, length: { maximum: 500 }
-  validate :year_from_before_year_to
+  validate :finish_at_before_start_at
```
Comment (psunix) [inline app/models/project.rb]: wrong order ... start_at_before_finish_at ....

**Comment by psunix** [comment]
latest-screenshot

this error still appears if date is invalid

**Comment by psunix** [comment]
latest-screenshot

this error still appears if date is invalid

File: db/migrate/20181120072351_change_year_to_date_range.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        revert_year_from(e)
+      end
+      remove_column table, :start_at
+      remove_column table, :finish_at
+    end
+  end
+
+  private
+
+  def update_start_at(entry)
```
Comment (psunix) [inline db/migrate/20181120072351_change_year_to_date_range.rb]: it's not only start_at ... finish_at too

File: spec/migrations/change_year_to_date_range_spec.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  context 'up' do
+
+    before { migration.down }
+
+    it 'changes columns to start_at and finish_at' do
+      migration.up
+
+      expect(project.finish_at).to eq(Date.parse('2012-01-01'))
+      expect(project.start_at).to eq(Date.parse('2015-12-01'))
+      expect(education.finish_at).to eq(Date.parse('2000-01-01'))
```
Comment (psunix) [inline spec/migrations/change_year_to_date_range_spec.rb]: should be set to December (12), shouldn't it ?

**Comment by psunix** [comment]
could you also add a accepntance test for the case: start_at after finish_at date ?

**Comment by Robin481** [comment]
Funktionell alles Top.

Offene Frage ist noch ob wir es (irgendwie) erlauben wollen keinen Monat anzugeben. Ich werde Ändu ansprechen.

**Comment by Robin481** [comment]
Bug:
Bei "from" Monat auswählen, e.g: 3
Speichern und dann wieder bearbeiten
Jahr bei "from" mit Pfeil um 1 erhöhen: Monat ändert von 3 auf 2.

**Pull request merged**